repository_service_tuf_worker/signer.py

```diff
@@ -1,51 +1,140 @@
+        ValueError: uri has no file name, or private key cannot be decoded,
+                or type does not match public key
+        cryptography.exceptions.UnsupportedAlgorithm: key type not supported
+
+    """
+
+    SCHEME = "fn"
+    DIR_VAR = "RSTUF_ONLINE_KEY_DIR"
+
+    @classmethod
+    def from_priv_key_uri(
+        cls,
+        priv_key_uri: str,
+        public_key: Key,
+        secrets_handler: Optional[SecretsHandler] = None,
+    ) -> "FileNameSigner":
+        _, _, file_name = priv_key_uri.partition(":")
+        if not file_name:
+            raise ValueError(
+                f"bad uri: expected '{cls.SCHEME}:<file name>', "
+                f"'got {priv_key_uri}'"
+            )
+
+        dir_ = os.environ[cls.DIR_VAR]
+
+        with open(Path(dir_, file_name), "rb") as f:
+            private_pem = f.read()
+
+        private_key = load_pem_private_key(private_pem, None)
+        return cls(private_key, public_key)
+
+
 RSTUF_ONLINE_KEY_URI_FIELD = "x-rstuf-online-key-uri"
 
 # Register non-default securesystemslib file signer
 # secure-systems-lab/securesystemslib#617
 SIGNER_FOR_URI_SCHEME[CryptoSigner.FILE_URI_SCHEME] = CryptoSigner
+# Register custom FileNameSigner
+SIGNER_FOR_URI_SCHEME[FileNameSigner.SCHEME] = FileNameSigner
+
+
+@contextmanager
+def isolated_env(env: dict[str, str]):
+    """Temporarily replace environment."""
+    orig_env = dict(os.environ)
+    os.environ.clear()
+    os.environ.update(env)
+    try:
+        yield
+    finally:
+        os.environ.clear()
+        os.environ.update(orig_env)
 
 
 class SignerStore:
-    """Generic signer store."""
+    """Generic signer store.
+
+    Parses ambient signer settings and provides them in an isolated environment
+    under the expected names.
+    """
 
     def __init__(self, settings: Dynaconf):
-        self._settings = settings
+        # Cache known ambient settings
+        self._ambient_settings: dict[str, str] = {}
+        if key_dir := settings.get("ONLINE_KEY_DIR"):
+            self._ambient_settings[FileNameSigner.DIR_VAR] = key_dir
+
+        # Cache KEYVAULT setting as fallback
+        self._vault = settings.get("KEYVAULT")
         self._signers: dict[str, Signer] = {}
 
     def get(self, key: Key) -> Signer:
         """Return signer for passed key.
 
         - signer is loaded from the uri included in the passed public key
           (see SIGNER_FOR_URI_SCHEME for available uri schemes)
+        - additional signer settings can be provided "ambiently" (see __init__)
         - RSTUF_KEYVAULT_BACKEND is used as fallback, if no URI is included
 
         """
 
         if key.keyid not in self._signers:
             if uri := key.unrecognized_fields.get(RSTUF_ONLINE_KEY_URI_FIELD):
-                self._signers[key.keyid] = Signer.from_priv_key_uri(uri, key)
+                # (Re-)export ambient settings in isolated environment
+                with isolated_env(self._ambient_settings):
```

Comment:
I wonder if we really need an isolated environment. I added it mostly to avoid weird conflicts between the `RSTUF_` prefixed config and the signer-specific expected environment variables.

Let's take AWS as example. The AWSSigner implementation in securesystemslib (used by #452) expects an environment variable `AWS_ENDPOINT_URL`. But we would likely ask the user to provide that value via `RSTUF_AWS_ENDPOINT_URL` setting.

Now, how should SignerStore behave if the `RSTUF_AWS_ENDPOINT_URL` setting is missing, but `AWS_ENDPOINT_URL` is defined? Unless we isolate (or vet) the environment, AWSSigner will happily accept the environment variable.

Reply:
A limitation of the isolated environment is that the signer can only access the ambient configuration when it is created (`from_priv_key_uri`) but not when it is used (`sign`).

This is not an issue for `FileNameSigner` or `AWSSigner` (tested in https://github.com/repository-service-tuf/repository-service-tuf-worker/commit/e1514aac496a5051afece6cad24fc3065bfedea5), but may be one for other Signer implementations.

Reply:
I suggest we move forward with the isolated environment and re-consider, if we need a Signer implementation where the does not work.